### PR TITLE
Feature/enforced and monitored routes

### PR DESCRIPTION
**Pull request by chen-zimmer-px**
The monitored and enforced routes by the new functionality.
The new functionality- 
**module mode is monitor:**
Enforced routes are specific routes that should be enforced. Other routes should be monitored.
**module mode is active_blocking:**
Monitored routes are specific routes that should be monitored. Other routes should be enforced.

### CODE REVIEW

File: src/main/java/com/perimeterx/api/verificationhandler/DefaultVerificationHandler.java [hunk 1]
```diff
@@ -50,7 +55,8 @@ public boolean handleVerification(PXContext context, HttpServletResponseWrapper
         }
         setPxhdCookie(context, responseWrapper);
         boolean shouldBypassMonitor = shouldBypassMonitor(context);
-        if ((pxConfiguration.getModuleMode().equals(ModuleMode.BLOCKING) || shouldBypassMonitor) && !verified) {
+
+        if (!verified && (context.isBlocking() || shouldBypassMonitor)) {
```
Comment (OmerBacharPX) [inline src/main/java/com/perimeterx/api/verificationhandler/DefaultVerificationHandler.java:59]: I know this is not your change but we always want to refactor thing to make them better - I think that `shouldBypassMonitor` should be combined with the current `isBlocking` function, at least calling `shouldBypassMonitor` from within `isBlocking`, this is all part of understanding if this request should be considered as monitored request or not.

File: src/main/java/com/perimeterx/models/PXContext.java [hunk 2]
```diff
@@ -316,20 +337,16 @@ private void setVidAndPxhd(Cookie[] cookies) {
...
 
-    public Boolean isSimulatedBlock(){
-        return this.simulatedBlock;
+    public Boolean isMonitoredRequest() {
+        return this.isMonitoredRequest;
     }
 
     public Boolean isBlocking() {
-        return !this.simulatedBlock;
+        return !this.isMonitoredRequest;
```
Comment (OmerBacharPX) [inline src/main/java/com/perimeterx/models/PXContext.java:349]: I think that isBlocking is a bad name, the `verified` variable already tells us if this request should be blocked or not (in general), the value that this function outputs represent if this should be treat as monitor or actual block when decided to be blocked, therefor the variable name is already good enough and we should use it instead.

`isBlocking` is used once other than your usage in `getRiskMode`, I checked this field in other enforcers and it should indicate if risk is in monitor or active blocking per request (exactly what `isMonitoredRequest` represents)

File: src/main/java/com/perimeterx/models/PXContext.java [hunk 3]
```diff
@@ -237,7 +237,7 @@ private void initContext(final HttpServletRequest request, PXConfiguration pxCon
         this.madeS2SApiCall = false;
         this.riskRtt = 0;
         this.httpMethod = request.getMethod();
-        this.simulatedBlock = pxConfiguration.getModuleMode().equals(ModuleMode.MONITOR);
+        this.isMonitoredRequest = checkIfMonitorRequest();
```
Comment (OmerBacharPX) [inline src/main/java/com/perimeterx/models/PXContext.java:240]: make sure that when sent on activities, this field is still called `simulated_block`

File: src/main/java/com/perimeterx/models/PXContext.java [hunk 4]
```diff
@@ -250,6 +250,27 @@ private void initContext(final HttpServletRequest request, PXConfiguration pxCon
...
+        Pattern pattern;
+        Matcher matcher;
+
+        for (String specialRoute : routes) {
+            pattern = Pattern.compile(specialRoute, Pattern.CASE_INSENSITIVE);
+            matcher = pattern.matcher(uri);
+            if (matcher.matches()) {
+                return true;
+            }
+        }
```
Comment (OmerBacharPX) [inline src/main/java/com/perimeterx/models/PXContext.java:263]: This is not clear enough, the configuration with the current implementation only accepts regex expressions, but what if the customer want exact path (we said the customer can also configure `"/profile"` and the Enforcer should treat this is special route only if the path is exactly /profile). In your current impl if the customer want to configure exact /profile he should add to the configuration `^/profile$` (start and ends with /profile).
Moreover, does this library accepts strings as input or it expects regex as input?

I think we should support both string (for exact paths) and regex as input but let's talk about it.

**Review by OmerBacharPX — CHANGES_REQUESTED** [review]

**Comment by ori-gold-px** [comment]
I would add to the CHANGELOG even now since you're changing this feature. I know we're going to release right after anyway, but it's good practice to make changes to the changelog on every feature add / bugfix so that all changes are accounted for with every PR and we don't have to dig up old history when doing a release.

File: pom.xml [hunk 5]
```diff
@@ -221,7 +221,7 @@
         <dependency>
             <groupId>org.projectlombok</groupId>
             <artifactId>lombok</artifactId>
-            <version>1.16.22</version>
+            <version>1.18.22</version>
```
Comment (ori-gold-px) [inline pom.xml:224]: Maybe also good to mention in the changelog!

**Review by ori-gold-px — CHANGES_REQUESTED** [review]

**Review by OmerBacharPX — APPROVED** [review]

**Review by ori-gold-px — APPROVED** [review]
Looks good!

**Pull request merged**